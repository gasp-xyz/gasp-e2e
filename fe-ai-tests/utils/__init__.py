# This file makes the utils directory a Python package
# This allows imports like: from utils.common_test_utils import x
