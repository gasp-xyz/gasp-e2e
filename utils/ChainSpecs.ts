import { BN_TEN, BN_ZERO } from "@mangata-finance/sdk";
import { BN, BN_HUNDRED, BN_THOUSAND } from "@polkadot/util";
import { ApiPromise } from "@polkadot/api";

export const TRANSFER_INSTRUCTIONS = 4;
export const WEIGHT_IN_SECONDS = new BN(1_000_000_000_000);

export interface ChainSpec {
  unitCostWeight: number;
  parachain: number;
  assets: Map<AssetSpec, AssetMeta>;
  foreign: Map<AssetSpec, AssetMeta>;
}

export interface AssetSpec {
  symbol: string;
  decimals: number;
  location: object;
  unit: BN;
}

interface AssetMeta {
  fps: BN;
  ed: BN;
  location: object;
}

export enum ChainId {
  Mg,
  Tur,
  Bifrost,
  Statemine,
}

export class AssetId {
  static Tur: AssetSpec = {
    symbol: "TUR",
    decimals: 10,
    location: { parents: 1, interior: { X1: { Parachain: 2114 } } },
    unit: BN_TEN.pow(new BN(10)),
  };
  static TurV3: AssetSpec = {
    symbol: "TUR",
    decimals: 10,
    location: { parents: 1, interior: { X1: { Parachain: 2114 } } },
    unit: BN_TEN.pow(new BN(10)),
  };

  static Mgx: AssetSpec = {
    symbol: "MGX",
    decimals: 18,
    location: {
      parents: 1,
      interior: { X2: [{ Parachain: 2110 }, { GeneralKey: "0x00000000" }] },
    },
    unit: BN_TEN.pow(new BN(18)),
  };
  static async getLocationFromChain(symbol: string, api: ApiPromise) {
    const entries = await api.query.assetRegistry.metadata.entries();
    const asset = entries.filter((entry) => {
      const elem = entry[1];
      const name = api.createType(
        "Vec<u8>",
        JSON.parse(JSON.stringify(elem)).symbol,
      );
      return name.toHuman() === symbol;
    })[0];
    return JSON.parse(JSON.stringify(asset[1].toHuman())).location;
  }
  static Bnc: AssetSpec = {
    symbol: "BNC",
    decimals: 12,
    location: {
      parents: 1,
      interior: { X2: [{ Parachain: 2001 }, { GeneralKey: "0x0001" }] },
    },
    unit: BN_TEN.pow(new BN(12)),
  };
  static ImbueBncV3: AssetSpec = {
    symbol: "BNC",
    decimals: 12,
    location: {
      parents: 0,
      interior: {
        X1: [
          {
            GeneralKey: {
              length: 2,
              data: "0x0001000000000000000000000000000000000000000000000000000000000000",
            },
          },
        ],
      },
    },
    unit: BN_TEN.pow(new BN(12)),
  };
  static BncV3: AssetSpec = {
    symbol: "BNC",
    decimals: 12,
    location: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: 2001 },
          {
            GeneralKey: {
              length: 2,
              data: "0x0001000000000000000000000000000000000000000000000000000000000000",
            },
          },
        ],
      },
    },
    unit: BN_TEN.pow(new BN(12)),
  };

  static USDt: AssetSpec = {
    symbol: "USDt",
    decimals: 6,
    location: {
      parents: 1,
      interior: {
        X3: [
          { Parachain: 1000 },
          { PalletInstance: 50 },
          { GeneralIndex: 1984 },
        ],
      },
    },
    unit: BN_TEN.pow(new BN(6)),
  };
}

export const ChainSpecs = new Map<ChainId, ChainSpec>([
  [
    ChainId.Tur,
    {
      unitCostWeight: 1_000_000_000,
      parachain: 2114,
      assets: new Map([
        [
          AssetId.Tur,
          {
            fps: new BN(416_000_000_000),
            ed: AssetId.Tur.unit.mul(BN_TEN),
            location: AssetId.Tur.location,
          },
        ],
      ]),
      foreign: new Map(),
    },
  ],
  [
    ChainId.Mg,
    {
      unitCostWeight: 150_000_000,
      parachain: 2110,
      assets: new Map(),
      foreign: new Map([
        [
          AssetId.Tur,
          {
            fps: new BN(537_600_000_000),
            ed: BN_ZERO,
            location: AssetId.Tur.location,
          },
        ],
        [
          AssetId.TurV3,
          {
            fps: new BN(537_600_000_000),
            ed: BN_ZERO,
            location: AssetId.TurV3.location,
          },
        ],
        [
          AssetId.Bnc,
          {
            fps: new BN(43008000000000),
            ed: BN_ZERO,
            location: AssetId.Bnc.location,
          },
        ],
        [
          AssetId.ImbueBncV3,
          {
            fps: new BN(43008000000000),
            ed: BN_ZERO,
            location: AssetId.BncV3.location,
          },
        ],
        [
          AssetId.BncV3,
          {
            fps: new BN(43008000000000),
            ed: BN_ZERO,
            location: AssetId.BncV3.location,
          },
        ],
        [
          AssetId.USDt,
          {
            fps: new BN(13440000),
            ed: BN_ZERO,
            location: AssetId.USDt.location,
          },
        ],
      ]),
    },
  ],
  [
    ChainId.Bifrost,
    {
      unitCostWeight: 200_000_000,
      parachain: 2001,
      assets: new Map([
        [
          AssetId.Bnc,
          {
            fps: new BN(9360000000000),
            ed: AssetId.Bnc.unit.div(BN_HUNDRED),
            location: {
              parents: 0,
              interior: { X1: { GeneralKey: "0x0001" } },
            },
          },
        ],
        [
          AssetId.BncV3,
          {
            fps: new BN(9360000000000),
            ed: AssetId.BncV3.unit.div(BN_HUNDRED),
            location: {
              parents: 0,
              interior: {
                X1: {
                  GeneralKey: {
                    length: 2,
                    data: "0x0001000000000000000000000000000000000000000000000000000000000000",
                  },
                },
              },
            },
          },
        ],
        [
          AssetId.ImbueBncV3,
          {
            fps: new BN(9360000000000),
            ed: AssetId.ImbueBncV3.unit.div(BN_HUNDRED),
            location: {
              parents: 0,
              interior: {
                X1: {
                  GeneralKey: {
                    length: 2,
                    data: "0x0001000000000000000000000000000000000000000000000000000000000000",
                  },
                },
              },
            },
          },
        ],
      ]),
      foreign: new Map(),
    },
  ],
  [
    ChainId.Statemine,
    {
      unitCostWeight: 1_000_000_000,
      parachain: 1000,
      assets: new Map([
        [
          AssetId.USDt,
          {
            fps: new BN(1),
            ed: AssetId.USDt.unit.div(BN_THOUSAND),
            location: {
              parents: 0,
              interior: {
                X2: [{ PalletInstance: 50 }, { GeneralIndex: 1984 }],
              },
            },
          },
        ],
      ]),
      foreign: new Map(),
    },
  ],
]);
