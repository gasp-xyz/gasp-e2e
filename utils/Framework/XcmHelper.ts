import {
  Blockchain,
  BuildBlockMode,
  setupWithServer,
} from "@acala-network/chopsticks";
import { StorageValues } from "@acala-network/chopsticks/lib/utils/set-storage";
import { ApiPromise, WsProvider } from "@polkadot/api";
import { HexString } from "@polkadot/util/types";
import { getPort } from "get-port-please";
import * as fs from "fs";
import { bufferToU8a, u8aToHex } from "@polkadot/util";
import { Assets } from "../Assets";
import { alice } from "../setup";
import { Sudo } from "../sudo";

export type SetupOption = {
  endpoint: string;
  blockNumber?: number;
  blockHash?: HexString;
  wasmOverride?: string;
  db?: string;
  types?: Record<string, any>;
  localPort?: number;
  buildBlockMode?: BuildBlockMode;
};

export type DevApi = {
  newBlock: (param?: { count?: number; to?: number }) => Promise<string>;
  setStorage: (values: StorageValues, blockHash?: string) => Promise<void>;
  timeTravel: (date: string | number) => Promise<number>;
  setHead: (hashOrNumber: string | number) => Promise<void>;
};

export type ApiContext = {
  uri: string;
  chain: Blockchain;
  ws: WsProvider;
  api: ApiPromise;
  dev: DevApi;
  teardown: () => Promise<void>;
};

export const setupContext = async ({
  endpoint,
  blockNumber,
  blockHash,
  wasmOverride,
  db,
  types,
  localPort,
  buildBlockMode,
}: SetupOption): Promise<ApiContext> => {
  // random port
  const port = localPort ? localPort : await getPort();
  const config = {
    endpoint,
    port,
    block: blockNumber || blockHash,
    mockSignatureHost: true,
    "build-block-mode": buildBlockMode ? buildBlockMode : BuildBlockMode.Manual,
    db,
    "wasm-override": wasmOverride,
    "registered-types": { types: types },
    "runtime-log-level": 5,
    runtimeLogLevel: 5,
  };
  const { chain, listenPort, close } = await setupWithServer(config);
  const uri = `ws://localhost:${listenPort}`;
  const ws = new WsProvider(uri);
  const api = await ApiPromise.create({
    provider: ws,
    types: types,
  });

  await api.isReady;

  return {
    uri,
    chain,
    ws,
    api,
    dev: {
      newBlock: (param?: { count?: number; to?: number }): Promise<string> => {
        return ws.send("dev_newBlock", [param]);
      },
      setStorage: (values: StorageValues, blockHash?: string) => {
        return ws.send("dev_setStorage", [values, blockHash]);
      },
      timeTravel: (date: string | number) => {
        return ws.send<number>("dev_timeTravel", [date]);
      },
      setHead: (hashOrNumber: string | number) => {
        return ws.send("dev_setHead", [hashOrNumber]);
      },
    },
    async teardown() {
      await api.disconnect();
      await close();
    },
  };
};
export async function upgradeMangata(mangata: ApiContext) {
  const path = `test/xcm/_releasesUT/0.30.0/mangata_kusama_runtime-0.30.0.RC.compact.compressed.wasm`;
  const wasmContent = fs.readFileSync(path);
  const hexHash = mangata.api!.registry.hash(bufferToU8a(wasmContent)).toHex();
  await Sudo.batchAsSudoFinalized(Assets.mintNative(alice));
  await Sudo.asSudoFinalized(
    Sudo.sudo(
      //@ts-ignore
      mangata.api!.tx.parachainSystem.authorizeUpgrade(hexHash)
    )
  );
  const wasmParam = Uint8Array.from(wasmContent);
  const hex = u8aToHex(wasmParam);
  const param = hex.toString();
  await mangata.api.tx.sudo
    .sudo(mangata.api.tx.parachainSystem.enactAuthorizedUpgrade(param))
    .signAndSend(alice.keyRingPair);

  await mangata.dev.newBlock();
  await mangata.dev.newBlock();
}
