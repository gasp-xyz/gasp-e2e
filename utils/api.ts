import { ApiPromise } from "@polkadot/api";
import { testLog } from "./Logger";
import { getEnvironmentRequiredVars, getMangataApiUrlPort } from "./utils";
import { Mangata, MangataInstance } from "@mangata-finance/sdk";
import { getPort } from "get-port-please";
import XcmNetworks from "./Framework/XcmNetworks";
import { BuildBlockMode } from "@acala-network/chopsticks";
import { ApiContext } from "./Framework/XcmHelper";

export let api: ApiPromise | null = null;
export let mangata: MangataInstance | null = null;
export let mangataChopstick: ApiContext | null = null;
export let chopstickUri: String;
export const getApi = () => {
  if (!api) {
    throw new Error("Api not initialized");
  }
  return api;
};

export const initApi = async (uri = "") => {
  if (!uri) {
    const { chainUri: envUri } = getEnvironmentRequiredVars();
    uri = envUri;
  }
  if (process.env.CHOPSTICK_ENABLED) {
    const mgaPort = getMangataApiUrlPort();
    const chopstickPort = await getPort();
    mangataChopstick = await XcmNetworks.mangata({
      localPort: chopstickPort,
      buildBlockMode: BuildBlockMode.Instant,
    });
    uri = uri.replace(mgaPort.toString(), chopstickPort.toString());
    chopstickUri = uri;
  }
  testLog.getLog().info(`TEST_INFO: Running test in ${uri}`);
  mangata = Mangata.instance([uri]);
  api = await mangata.api();
  return api;
};

export async function getMangataInstance(uri = ""): Promise<MangataInstance> {
  if (!api) {
    await initApi(uri);
  }
  return mangata!;
}
export const disconnect = async () => {
  //TODO:sdkV2
  //  if (mangata) {
  //    await mangata.disconnect();
  //  } else {
  //    await Mangata.instance([uri]).disconnect();
  //  }
};
