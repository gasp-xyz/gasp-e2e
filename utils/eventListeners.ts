/* eslint-disable no-loop-func */
import { MangataGenericEvent } from "@mangata-finance/sdk";
import { ApiPromise } from "@polkadot/api";
import { BN } from "@polkadot/util";
import _, { reject } from "lodash-es";
import { getApi, getMangataInstance } from "./api";
import { logEvent, testLog } from "./Logger";
import { api } from "./setup";
import { getEventErrorFromSudo } from "./txHandler";
import { User } from "./User";
import { getEnvironmentRequiredVars } from "./utils";
import { Codec } from "@polkadot/types/types";

// lets create a enum for different status.
export enum ExtrinsicResult {
  ExtrinsicSuccess,
  ExtrinsicFailed,
  ExtrinsicUndefined,
}

///Class that stores the event result.
export class EventResult {
  /**
   *
   */
  constructor(
    state: ExtrinsicResult = ExtrinsicResult.ExtrinsicUndefined,
    data: any,
  ) {
    this.state = state;
    this.data = data;
  }

  state: ExtrinsicResult;
  data: String;
}

// for testing
export const getEventResult = (
  section: any,
  method: any,
  module_index: any,
) => {
  const api = getApi();

  return new Promise<EventResult>(async (resolve) => {
    const unsubscribe = (await api.query.system.events((events: any) => {
      events.forEach((record: any) => {
        const { event } = record;
        if (event.section === section && event.method === method) {
          unsubscribe();
          resolve(
            new EventResult(
              ExtrinsicResult.ExtrinsicSuccess,
              JSON.parse(event.data.toString()),
            ),
          );
        } else if (
          event.section === "system" &&
          event.method === "ExtrinsicFailed" &&
          (JSON.parse(event.data.toString())[0].Module.index = module_index)
        ) {
          unsubscribe();
          resolve(
            new EventResult(
              ExtrinsicResult.ExtrinsicFailed,
              JSON.parse(event.data.toString())[0].Module.error,
            ),
          );
        }
      });
    })) as any;
  });
};

export const waitNewBlock = () => {
  const api = getApi();
  let count = 0;
  return new Promise(async (resolve) => {
    const unsubscribe = await api.rpc.chain.subscribeNewHeads((header: any) => {
      if (++count === 2) {
        testLog.getLog().info(`Chain is at block: #${header.number}`);
        unsubscribe();
        resolve(true);
      }
    });
  });
};

export function filterEventData(
  result: MangataGenericEvent[],
  method: string,
): any[] {
  return result
    .filter((event) => `${event.section}.${event.method}` === method)
    .map((event) => event.event.toHuman().data);
}

export function findEventData(result: MangataGenericEvent[], method: string) {
  return filterEventData(result, method)[0];
}

export async function waitSudoOperationSuccess(
  checkingEvent: MangataGenericEvent[],
) {
  const filterBootstrapEvent = checkingEvent.filter(
    (extrinsicResult) => extrinsicResult.method === "Sudid",
  );

  const userBootstrapCall = filterBootstrapEvent[0].event.data[0].toString();

  expect(userBootstrapCall).toContain("Ok");
}

export async function waitSudoOperationFail(
  checkingEvent: MangataGenericEvent[],
  expectedErrors: string[],
) {
  const filterBootstrapEvent = checkingEvent.filter(
    (extrinsicResult) => extrinsicResult.method === "Sudid",
  );

  const BootstrapError = await getEventErrorFromSudo(filterBootstrapEvent);

  expect(expectedErrors).toContain(BootstrapError.method);
}

export const waitForEvents = async (
  api: ApiPromise,
  method: string,
  blocks: number = 10,
  withData: string = "",
): Promise<CodecOrArray> => {
  return new Promise(async (resolve, reject) => {
    let counter = 0;
    const unsub = await api.rpc.chain.subscribeFinalizedHeads(async (head) => {
      const events = await (await api.at(head.hash)).query.system.events();
      counter++;
      testLog
        .getLog()
        .info(
          `â†’ find on ${api.runtimeChain} for '${method}' event, attempt ${counter}, head ${head.hash}`,
        );

      events.forEach((e) => logEvent(api.runtimeChain, e));

      const filtered = _.filter(
        events,
        ({ event }) =>
          `${event.section}.${event.method}` === method &&
          (withData.length > 0
            ? JSON.stringify(event.data.toHuman()).includes(withData)
            : true),
      );
      if (filtered.length > 0) {
        resolve(filtered);
        unsub();
      }
      if (counter === blocks) {
        reject(`method ${method} not found within blocks limit`);
      }
    });
  });
};

export const waitForRewards = async (
  user: User,
  liquidityAssetId: BN,
  max: number = 20,
) =>
  new Promise(async (resolve) => {
    let numblocks = max;
    const unsub = await api.rpc.chain.subscribeNewHeads(async (header) => {
      numblocks--;
      const { chainUri } = getEnvironmentRequiredVars();
      const mangata = await getMangataInstance(chainUri);
      const price = await mangata.rpc.calculateRewardsAmount({
        address: user.keyRingPair.address,
        liquidityTokenId: liquidityAssetId.toString(),
      });

      if (price.gtn(0)) {
        unsub();
        resolve({});
      } else {
        testLog
          .getLog()
          .info(
            `#${header.number}  ${user.keyRingPair.address} (LP${liquidityAssetId}) - no rewards yet`,
          );
      }
      if (numblocks < 0) {
        unsub();
        reject(
          `Waited too long for rewards :( #${header.number}  ${user.keyRingPair.address} (LP${liquidityAssetId} `,
        );
      }
    });
  });

type CodecOrArray = Codec | Codec[];

const processCodecOrArray = (codec: CodecOrArray, fn: (c: Codec) => any) =>
  Array.isArray(codec) ? codec.map(fn) : fn(codec);

const toHuman = (codec: CodecOrArray) =>
  processCodecOrArray(codec, (c) => c?.toHuman?.() ?? c);
const toJson = (codec: CodecOrArray) =>
  processCodecOrArray(codec, (c) => c?.toJSON?.() ?? c);
const toHex = (codec: CodecOrArray) =>
  processCodecOrArray(codec, (c) => c?.toHex?.() ?? c);

export const matchSnapshot = (
  codec: CodecOrArray | Promise<CodecOrArray>,
  message?: string,
) => {
  return expect(Promise.resolve(codec).then(toHuman)).resolves.toMatchSnapshot(
    message,
  );
};

export const expectEvent = (codec: CodecOrArray, event: any) => {
  return expect(toHuman(codec)).toEqual(
    expect.arrayContaining([expect.objectContaining(event)]),
  );
};

export const expectHuman = (codec: CodecOrArray) => {
  return expect(toHuman(codec));
};

export const expectJson = (codec: CodecOrArray) => {
  return expect(toJson(codec));
};

export const expectHex = (codec: CodecOrArray) => {
  return expect(toHex(codec));
};

type EventFilter = string | { method: string; section: string };

const _matchEvents = async (
  msg: string,
  events: Promise<Codec[] | Codec>,
  ...filters: EventFilter[]
) => {
  let data = toHuman(await events).map(
    ({ event: { index: _, ...event } }: any) => event,
  );
  if (filters) {
    const filtersArr = Array.isArray(filters) ? filters : [filters];
    data = data.filter((evt: any) => {
      return filtersArr.some((filter) => {
        if (typeof filter === "string") {
          return evt.section === filter;
        }
        const { section, method } = filter;
        return evt.section === section && evt.method === method;
      });
    });
  }
  return expect(data).toMatchSnapshot(msg);
};

export const matchEvents = async (
  events: Promise<Codec[] | Codec> | Codec[],
  ...filters: EventFilter[]
) => {
  return _matchEvents("events", redact(events), ...filters);
};

export const matchSystemEvents = async (
  { api }: { api: ApiPromise },
  ...filters: EventFilter[]
) => {
  await _matchEvents(
    "system events",
    redact(api.query.system.events()),
    ...filters,
  );
};

export const matchUmp = async ({ api }: { api: ApiPromise }) => {
  expect(await api.query.parachainSystem.upwardMessages()).toMatchSnapshot(
    "ump",
  );
};

export const redact = async (data: any | Promise<any>) => {
  const json = toHuman(await data);

  const process = (obj: any): any => {
    if (obj == null) {
      return obj;
    }
    if (Array.isArray(obj)) {
      return obj.map(process);
    }
    if (typeof obj === "number") {
      return "(redacted)";
    }
    if (typeof obj === "string") {
      if (obj.match(/^[\d,]+$/) || obj.match(/0x[0-9a-f]{64}/)) {
        return "(redacted)";
      }
      return obj;
    }
    if (typeof obj === "object") {
      return Object.fromEntries(
        Object.entries(obj).map(([k, v]) => [k, process(v)]),
      );
    }
    return obj;
  };

  return process(json);
};

export const expectExtrinsicSuccess = (events: Codec[]) => {
  expectEvent(events, {
    event: expect.objectContaining({
      method: "ExtrinsicSuccess",
      section: "system",
    }),
  });
};
export const expectMGAExtrinsicSuDidSuccess = (
  events: MangataGenericEvent[],
) => {
  const anyError = events
    .filter((x) => x.method === "SudoAsDone" && x.section === "sudo")
    .some((x) => JSON.parse(JSON.stringify(x.eventData[0].data)).err);
  if (!anyError) {
    testLog.getLog().warn(JSON.stringify(events));
  }
  expect(anyError).toBeFalsy();
  return events!;
};
export const expectMGAExtrinsicSuDidFailed = (
  events: MangataGenericEvent[],
) => {
  const sudoErrorEvent = events.find(
    (x) =>
      x.method === "SudoAsDone" &&
      x.section === "sudo" &&
      JSON.parse(JSON.stringify(x.eventData[0].data)).err !== undefined,
  );
  expect(sudoErrorEvent).not.toBeNull();
  return sudoErrorEvent!;
};
