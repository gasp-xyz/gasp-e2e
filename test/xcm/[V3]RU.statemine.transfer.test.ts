import { connectParachains } from "@acala-network/chopsticks";
import { KeyringPair } from "@polkadot/keyring/types";
import { BN_THOUSAND, BN } from "@polkadot/util";
import { AssetId } from "../../utils/ChainSpecs";
import { ApiContext } from "../../utils/Framework/XcmHelper";
import XcmNetworks from "../../utils/Framework/XcmNetworks";
import { devTestingPairs, setupApi, setupUsers } from "../../utils/setup";
import {
  expectExtrinsicSuccess,
  expectJson,
  matchSystemEvents,
} from "../../utils/validators";
import { mangataChopstick } from "../../utils/api";
import { BN_BILLION, Mangata } from "@mangata-finance/sdk";
import { Codec } from "@polkadot/types/types";
import { expectEvent, matchEvents } from "../../utils/eventListeners";

/**
 * @group xcm
 * @group proxied
 */
describe("XCM tests for Mangata <-> Statemine", () => {
  let statemine: ApiContext;
  let mangata: ApiContext;
  let alice: KeyringPair;

  beforeAll(async () => {
    await setupApi();
    await setupUsers();
    statemine = await XcmNetworks.statemine();
    mangata = mangataChopstick!;
    await connectParachains([statemine.chain, mangata.chain]);
    alice = devTestingPairs().alice;
    await mangata.dev.setStorage({
      Tokens: {
        Accounts: [
          [[alice.address, { token: 30 }], { free: 1000e6 }],
          [[alice.address, { token: 31 }], { free: 1000e10 }],
          [
            [alice.address, { token: 0 }],
            { free: AssetId.Mgx.unit.mul(BN_BILLION).toString() },
          ],
        ],
      },
      Sudo: {
        Key: alice.address,
      },
    });
    // await upgradeMangata(mangata);
  });

  beforeEach(async () => {
    await mangata.dev.setStorage({
      Tokens: {
        Accounts: [
          [[alice.address, { token: 30 }], { free: 1000e6 }],
          [[alice.address, { token: 31 }], { free: 1000e10 }],
          [
            [alice.address, { token: 0 }],
            { free: AssetId.Mgx.unit.mul(BN_THOUSAND).toString() },
          ],
        ],
      },
      Sudo: {
        Key: alice.address,
      },
    });
    await statemine.dev.setStorage({
      System: {
        Account: [[[alice.address], { providers: 1, data: { free: 10e12 } }]],
      },
      Assets: {
        Account: [
          [[1984, alice.address], { balance: 1000e6 }],
          [[8, alice.address], { balance: 1000e10 }],
        ],
      },
    });
  });

  it("mangata transfer assets to statemine", async () => {
    const mgaSdk = Mangata.instance([mangata.uri]);
    await mgaSdk.xTokens.withdraw({
      account: alice,
      amount: new BN(10e6),
      destinationAddress: alice.address,
      parachainId: 1000,
      tokenSymbol: "USDT",
      withWeight: "Unlimited",
      txOptions: {
        async extrinsicStatus(events) {
          expectExtrinsicSuccess(events as any[] as Codec[]);
          expectEvent(events as any[] as Codec[], {
            event: expect.objectContaining({
              section: "xTokens",
              method: "TransferredMultiAssets",
            }),
          });
        },
      },
    });
    await mangata.chain.newBlock();

    await statemine.chain.newBlock();
    expectJson(
      await mangata.api.query.tokens.accounts(alice.address, 30),
    ).toMatchSnapshot();

    expect(
      await statemine.api.query.assets.account(1984, alice.address),
    ).toMatchSnapshot();

    await matchSystemEvents(mangata, "xcmpQueue", "Success");
  });

  it("statemine transfer assets to mangata", async () => {
    const mgaSdk = Mangata.instance([mangata.uri]);
    await mgaSdk.xTokens.depositFromStatemine({
      account: alice,
      assets: {
        V3: [
          {
            id: {
              Concrete: {
                parents: 0,
                interior: {
                  X2: [{ PalletInstance: 50 }, { GeneralIndex: 1984 }],
                },
              },
            },
            fun: { Fungible: 10e6 },
          },
        ],
      },
      beneficiary: {
        V3: {
          parents: 0,
          interior: {
            X1: {
              AccountId32: {
                network: undefined,
                id: alice.addressRaw,
              },
            },
          },
        },
      },
      destination: {
        V3: {
          parents: 1,
          interior: {
            X1: { Parachain: 2110 },
          },
        },
      },
      feeAssetItem: 0,
      url: statemine.uri,
      weightLimit: "Unlimited",
      txOptions: {
        async extrinsicStatus(events) {
          await matchEvents(events as any as Codec[], "polkadotXcm");
        },
      },
    });

    /**
    const tx = await sendTransaction(
      statemine.api.tx.polkadotXcm
        .limitedReserveTransferAssets(
          {
            V3: {
              parents: 1,
              interior: {
                X1: { Parachain: 2110 },
              },
            },
          },
          {
            V3: {
              parents: 0,
              interior: {
                X1: {
                  AccountId32: {
                    network: undefined,
                    id: alice.addressRaw,
                  },
                },
              },
            },
          },
          {
            V3: [
              {
                id: {
                  Concrete: {
                    parents: 0,
                    interior: {
                      X2: [{ PalletInstance: 50 }, { GeneralIndex: 1984 }],
                    },
                  },
                },
                fun: { Fungible: 10e6 },
              },
            ],
          },
          0,
          "Unlimited"
        )
        .signAsync(alice, { nonce: 0 })
    );
     */
    await statemine.chain.newBlock();

    //  await matchEvents(tx.events, "polkadotXcm");

    expect(
      await statemine.api.query.assets.account(1984, alice.address),
    ).toMatchSnapshot();

    await mangata.chain.newBlock();

    expectJson(
      await mangata.api.query.tokens.accounts(alice.address, 30),
    ).toMatchSnapshot();

    await matchSystemEvents(mangata, "xcmpQueue", "Success");
  });
  it("[RMRK] mangata transfer assets to statemine", async () => {
    const mgaSdk = Mangata.instance([mangata.uri]);
    await mgaSdk.xTokens.withdraw({
      account: alice,
      amount: new BN(10e10),
      destinationAddress: alice.address,
      parachainId: 1000,
      tokenSymbol: "RMRK",
      withWeight: "Unlimited",
      txOptions: {
        async extrinsicStatus(events) {
          expectExtrinsicSuccess(events as any[] as Codec[]);
          expectEvent(events as any[] as Codec[], {
            event: expect.objectContaining({
              section: "xTokens",
              method: "TransferredMultiAssets",
            }),
          });
        },
      },
    });
    await mangata.chain.newBlock();

    await statemine.chain.newBlock();
    expectJson(
      await mangata.api.query.tokens.accounts(alice.address, 31),
    ).toMatchSnapshot();

    expect(
      await statemine.api.query.assets.account(8, alice.address),
    ).toMatchSnapshot();

    await matchSystemEvents(mangata, "xcmpQueue", "Success");
  });

  it("[RMRK] statemine transfer assets to mangata", async () => {
    const mgaSdk = Mangata.instance([mangata.uri]);
    await mgaSdk.xTokens.depositFromStatemine({
      account: alice,
      assets: {
        V3: [
          {
            id: {
              Concrete: {
                parents: 0,
                interior: {
                  X2: [{ PalletInstance: 50 }, { GeneralIndex: 8 }],
                },
              },
            },
            fun: { Fungible: 10e10 },
          },
        ],
      },
      beneficiary: {
        V3: {
          parents: 0,
          interior: {
            X1: {
              AccountId32: {
                network: undefined,
                id: alice.addressRaw,
              },
            },
          },
        },
      },
      destination: {
        V3: {
          parents: 1,
          interior: {
            X1: { Parachain: 2110 },
          },
        },
      },
      feeAssetItem: 0,
      url: statemine.uri,
      weightLimit: "Unlimited",
      txOptions: {
        async extrinsicStatus(events) {
          await matchEvents(events as any as Codec[], "polkadotXcm");
        },
      },
    });
    await statemine.chain.newBlock();

    expect(
      await statemine.api.query.assets.account(8, alice.address),
    ).toMatchSnapshot();

    await mangata.chain.newBlock();

    expectJson(
      await mangata.api.query.tokens.accounts(alice.address, 31),
    ).toMatchSnapshot();

    await matchSystemEvents(mangata, "xcmpQueue", "Success");
  });
});
