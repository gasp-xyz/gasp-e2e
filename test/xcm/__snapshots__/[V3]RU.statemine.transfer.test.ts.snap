// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XCM tests for Mangata <-> Statemine [RMRK] mangata transfer assets to statemine 1`] = `
{
  "free": 9900000000000,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`XCM tests for Mangata <-> Statemine [RMRK] mangata transfer assets to statemine 2`] = `
{
  "balance": 10099999281400,
  "extra": null,
  "reason": {
    "consumer": null,
  },
  "status": "Liquid",
}
`;

exports[`XCM tests for Mangata <-> Statemine [RMRK] mangata transfer assets to statemine: system events 1`] = `[]`;

exports[`XCM tests for Mangata <-> Statemine [RMRK] statemine transfer assets to mangata 1`] = `
{
  "balance": 9900000000000,
  "extra": null,
  "reason": {
    "consumer": null,
  },
  "status": "Liquid",
}
`;

exports[`XCM tests for Mangata <-> Statemine [RMRK] statemine transfer assets to mangata 2`] = `
{
  "free": 10099935488000,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`XCM tests for Mangata <-> Statemine [RMRK] statemine transfer assets to mangata: system events 1`] = `[]`;

exports[`XCM tests for Mangata <-> Statemine mangata transfer assets to statemine 1`] = `
{
  "free": 990000000,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`XCM tests for Mangata <-> Statemine mangata transfer assets to statemine 2`] = `
{
  "balance": 1009992814,
  "extra": null,
  "reason": {
    "consumer": null,
  },
  "status": "Liquid",
}
`;

exports[`XCM tests for Mangata <-> Statemine mangata transfer assets to statemine: system events 1`] = `[]`;

exports[`XCM tests for Mangata <-> Statemine statemine transfer assets to mangata 1`] = `
{
  "balance": 990000000,
  "extra": null,
  "reason": {
    "consumer": null,
  },
  "status": "Liquid",
}
`;

exports[`XCM tests for Mangata <-> Statemine statemine transfer assets to mangata 2`] = `
{
  "free": 1009989920,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`XCM tests for Mangata <-> Statemine statemine transfer assets to mangata: system events 1`] = `[]`;
